Charm documentation: readme, reference, explanation, tutorials and how to contribute (#43)
canonical/jenkins-agent-k8s-operator: docs/explanation/charm-architecture.md

# Charm architecture

> At its core, [Jenkins agents](https://www.jenkins.io/doc/book/managing/nodes/#components-of-distributed-builds) are [Java](https://www.java.com/en/) applications executing the Jobs in behalf of [Jenkins](https://www.jenkins.io/) itself.

Author: Oops. Thx

Pebble is a lightweight, API-driven process supervisor that is responsible for configuring processes to run in a container and controlling those processes throughout the workload lifecycle.

Pebble `services` are configured through [layers](https://github.com/canonical/pebble#layer-specification), and the following containers represent each one a layer forming the effective Pebble configuration, or `plan`:

1. An [Jenkins agent](https://www.jenkins.io/doc/book/managing/nodes/#components-of-distributed-builds) container, which manages the CI jobs.

As a result, if you run a `kubectl get pods` on a namespace named for the Juju model you've deployed the Jenkins agent k8s charm into, you'll see something like the following:

```bash
NAME                             READY   STATUS            RESTARTS   AGE
jenkins-agent-k8s-0              2/2     Running           0          2m2s
```

This shows there are 2 containers - the Jenkins agent one and container for the charm code itself.

> And if you run `kubectl describe pod jenkins-agent-k8s-0`, all the containers will have as Command ```/charm/bin/pebble```. That's because Pebble is responsible for the processes startup as explained above.

Reviewer: ```suggestion
If you run `kubectl describe pod jenkins-agent-k8s-0`, all the containers will `/charm/bin/pebble` as the entrypoint command. That's because Pebble is responsible for the processes startup as explained above.
```

## OCI images

We use [Rockcraft](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) to build the OCI Image for the Jenkins agent. 
The image is defined in the [Jenkins agent k8s ROCK](https://github.com/canonical/jenkins-agent-k8s-operator/blob/main/jenkins_agent_k8s_rock/).
They are published to [Charmhub](https://charmhub.io/), the official repository of charms.
This is done by publishing a resource to Charmhub as described in the [Juju SDK How-to guides](https://juju.is/docs/sdk/publishing).

## Containers

> Configuration files for the containers can be found in the respective directories that define the ROCKs, see the section above.

Reviewer: ```suggestion
Configuration files for the containers can be found in the respective directories that define the ROCKs. See the section above.
```
